src/Build/Definition/Project.cs

```diff
@@ -1117,32 +1129,161 @@ public List<GlobResult> GetAllGlobs(string itemType)
             return GetAllGlobs(GetItemElementsByType(GetEvaluatedItemElements(), itemType));
         }
 
+        private struct CumulatedRemoveElementData
+        {
+            public ImmutableList<IMSBuildGlob>.Builder Globs { get; set; }
+            public ImmutableHashSet<string>.Builder FragmentStrings { get; set; }
+        }
+
         private List<GlobResult> GetAllGlobs(List<ProjectItemElement> projectItemElements)
         {
-            return projectItemElements
-                .AsParallel()
-                .AsOrdered()
-                .SelectMany(GetAllGlobs)
-                .ToList();
+            if (projectItemElements.Count == 0)
+            {
+                return new List<GlobResult>();
+            }
+
+            // scan the project elements in reverse order and build globbing information for each include element
+            // based on the fact that relevant removes for a particular include element (element A) consist of:
+            // - all the removes seen by the next include statement of A's type (element B which appears after A in file order)
+            // - new removes between A and B
```

Comment:
Removes that apply to A but not to B. Spacially, these are placed between A's element and B's element.

```xml
1. <A Include="A"/>
2. <A Remove="..."/> // this remove applies to the A includes
3. <A Include="B"/>
4. <A Remove="..."/> // this remove applies to the A and B includes
5. <A Include="C"/>
6. <A Remove="..."/> // this remove applies to the A, B, and C includes

```

So A's applicable removes are composed of:
- all the removes seen by the next include statement of A's type (element B which appears after A in file order). In this example that's Removes 4 and 6.
- new removes between A and B. In this example that's Remove 2.